### PR TITLE
un-html, formatting, adding text from discussions and drafts

### PR DESCRIPTION
**Pull request by dericed**
@epiil and I started to integrate text from email discussions to the specification. I also did a lot of work to remove html formatting and other formatting that unlike other RFC documents. Because these patches integrate some of the discussion and draft with the specification there are now a few internal contradictions (such as the comments about the discrepancies in notes about the possible number of Element IDs). I can file issues for those or continue to discuss on the listserv.

### CODE REVIEW

File: specification.markdown [hunk 1]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
+encoding. The variable size integer begins with zero or more zero bits
+to define the width of the integer. Zero zeroes means a width of one
+byte, one zero a width of two bytes etc. The zeroes are followed by a
+marker of one set bit and then follows the actual integer value. The
+integer data consists of alignment data and tail data. The alignment
+data together with the width descriptor and the marker makes up one ore
+more complete bytes. The tail data is as many bytes as there were
+zeroes in the width descriptor, i.e. width-1.
+
+VINT           = VINT_WIDTH VINT_MARKER VINT_DATA
```
Comment (robUx4) [inline specification.markdown]: There's a formatting issue on the markdown rendering

File: specification.markdown [hunk 2]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
- </tbody>
-</table>
+## Structure
+
+Syntactically an EBML document is a list of EBML elements. Each element
+has three parts; an Element ID, an Element Data Size and the element
+data payload. The element data payload is then either data, implicitly
+typed by the Element ID, or a list of EBML elements.
+
+EBML       = *ELEMENT
```
Comment (robUx4) [inline specification.markdown]: There's a formatting issue on the Markdown rendering

File: specification.markdown [hunk 3]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
+ELEMENT_ID = VINT
+SIZE       = VINT
+
+EBML uses big endian/network order byte order, i.e. most significant
+bit first. All of the tokens above are byte aligned.
+
+## Variable Size Integer
+
+For both Element ID and Element Data Size EBML uses a variable size
+integer, coded according to a schema similar to that of UTF-8 [UTF-8]
```
Comment (robUx4) [inline specification.markdown]: the `[UTF-8]` should probably link somewhere or is redundant ?

File: specification.markdown [hunk 4]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
+    Multiple:       No
+    Range:          -
+    Default:        -
+    Element Type:   Binary
+    Description:    Used to void damaged data, to avoid unexpected
+                    behaviors when using damaged data. The content is
+                    discarded. Also used to reserve space in a
+                    sub-element for later use.
+
+### signature
```
Comment (robUx4) [inline specification.markdown]: Should mention that it's deprecated.

**Comment by robUx4** [comment]
As a general remark, avoid more than one commit per pull request. Git tools allow to turn several commits into one. You'll have a harder time editing your pull request when you have to edit the 1st of a batch of 37 commits.

Since it's the early work and a lot of changes are needed it's fine. The detail of why a change at a specific line/word won't matter much.

File: specification.markdown (hunk 1 above)
Comment (dericed) [inline specification.markdown]: Cleaned up in https://github.com/MediaArea/ebml-specification/commit/ca14fd2b9ec38cc4cae9b8730ff89a3f97160e6c

File: specification.markdown (hunk 2 above)
Comment (dericed) [inline specification.markdown]: fixed in https://github.com/MediaArea/ebml-specification/commit/a3ae17f4d8e77295ff899a2b7a4a0923c6105e12

File: specification.markdown (hunk 1 above)
Comment (dericed) [inline specification.markdown]: ah, I misunderstood what you were talking about at first. New patch to format the list: https://github.com/MediaArea/ebml-specification/commit/a3ae17f4d8e77295ff899a2b7a4a0923c6105e12

File: specification.markdown (hunk 4 above)
Comment (dericed) [inline specification.markdown]: @robUx4 Is the signature section deprecated or simply in draft form and never finalized?

**Comment by dericed** [comment]
any further pending issues on this PR?

File: specification.markdown (hunk 4 above)
Comment (robUx4) [inline specification.markdown]: I'd say in deprecated. libebml doesn't handle it. So not sure if anyone ever implemented support.
A signature is good if it's external to the content. Or, as in PGP, to prove the writer is the write person, but then it requires some key exchange. To verify data are not altered we already have CRC. Beyond that it's overkill with no added security.

File: specification.markdown [hunk 5]
```diff
@@ -1,66 +1,140 @@
...
+
+This document describes the EBML binary syntax, its semantic
+interpretation and the syntax of a textual document type definition
+format used to define the rules and constraints of an EBML language.
+
+## Notations and Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in RFC 2119.
```
Comment (robUx4) [inline specification.markdown]: Link to the RFC needed

**Comment by mbunkus** [comment]
@dericed Are you continuing to squash commits? Or should we merge this?

File: specification.markdown [hunk 6]
```diff
@@ -1,66 +1,140 @@
...
-    precise beginning of the millennium (at
-    2001-01-01T00:00:00,000000000 UTC)
--   master-element - contains other EBML sub-elements of the next lower
-    level
--   Binary - not interpreted by the parser
+Data Type        | Definition
+-----------------|----------------------------------------------------
+Signed Integer   | Big-endian, any size from 1 to 8 octets
+Unsigned Integer | Big-endian, any size from 1 to 8 octets
+Float            | Big-endian, defined for 4 and 8 octets (32, 64
```
Comment (robUx4) [inline specification.markdown]: Do we need the line formatting here ? We could at least have a 80 chars policy. This one ends at 66.

Comment (mbunkus) [inline specification.markdown]: I slightly prefer a tabular layout as it seems to be a bit easier to read, but I do agree that the lines are awfully short.

File: specification.markdown [hunk 7]
```diff
@@ -1,66 +1,140 @@
...
+Binary           | not interpreted by the parser
 
 As well as defining standard data types, EBML uses a system of Elements
-to make up an EBML "document." Elements incorporate an Element ID, a
-descriptor for the size of the element, and the binary data itself.
-Futher, Elements can be nested, or contain, Elements of a lower "level."
+to make up an EBML "document". Elements incorporate an Element ID, a
+descriptor for the size of the element (known as the Element Data
+Size), and the binary data itself. Futher, Elements can be nested or
+contain Elements of a lower "level".
```
Comment (robUx4) [inline specification.markdown]: I think the "lower" level is confusing, given in the specs we increment the number for nested elements. But that can be another pull request.

File: specification.markdown [hunk 8]
```diff
@@ -1,66 +1,140 @@
...
+default, widths up to 4. Another maximum width value can be set by
+setting another value to EBMLMaxIDLength in the EBML header. See
+section 5.1. Element IDs are always encoded in their shortest form, e.g. 1 is always encoded as 0x81 and never as 0x4001. This limits the number of Element IDs in every class with the number of Element IDs in the previous classes. Furthermore, values with all data bits set to 1 and all data bits set to 0 are reserved, hence the effective number of Element IDs are as follows for different widths. Note that the shortest encoding form for 127 is 0x407f since 0x7f is reserved.
+
+Width | Class ID | Number of Element IDs
+------|----------|------------------------
+1     | Class A  | 2^7-2     = 126
+2     | Class B  | 2^14-2^7  = 16,256
+3     | Class C  | 2^21-2^14 = 2,080,768
+4     | Class D  | 2^28-2^21 = 266,338,304
```
Comment (mbunkus) [inline specification.markdown]: I just noticed this. We were talking about »Class-ID« and »Class ID« and »Element ID« and that all those terms are confusing. I don't see a particular need for »Class ID« here and would prefer »Class name« as the column header. That would cut down on the number of uses of »xyz ID«.

File: specification.markdown [hunk 9]
```diff
@@ -1,66 +1,140 @@
...
+
 
-Element IDs are outlined as follows, beginning with the ID itself,
-followed by the Data Size, and then the non-interpreted Binary itself:
+### Element ID
+
+The EBML Element ID is encoded as a variable size integer with, by
+default, widths up to 4. Another maximum width value can be set by
+setting another value to EBMLMaxIDLength in the EBML header. See
+section 5.1. Element IDs are always encoded in their shortest form, e.g. 1 is always encoded as 0x81 and never as 0x4001. This limits the number of Element IDs in every class with the number of Element IDs in the previous classes. Furthermore, values with all data bits set to 1 and all data bits set to 0 are reserved, hence the effective number of Element IDs are as follows for different widths. Note that the shortest encoding form for 127 is 0x407f since 0x7f is reserved.
```
Comment (robUx4) [inline specification.markdown]: This limitation to the shortest form was not in the original specs. I think libebml doesn't have this limitation.
This should be part of another pull request as it's subject to debate.

File: specification.markdown [hunk 10]
```diff
@@ -1,66 +1,140 @@
...
-        Length Descriptor.
-    -   Any ID where all x's are composed entirely of 1's is a Reserved
-        ID, thus the -1 in the definitions above.
-    -   The Reserved IDs (all x set to 1) are the only IDs that may
-        change the Length Descriptor.
+        length of the Element ID. The number of leading 0's + 1 is the
+        length of the ID in octets. We will refer to the leading bits
+        as the Length Descriptor.
+    -   Any Element ID where all x's are composed entirely of 1's is a
+        Reserved Element ID, thus the -1 in the definitions above.
```
Comment (robUx4) [inline specification.markdown]: It's actually -2 above.

File: specification.markdown (hunk 9 above)
Comment (mbunkus) [inline specification.markdown]: It's coming from the RFC; neither the old EBML specs nor the Matroska specs contain such a limiation. So yes, we should move that to a separate pull request, please, as backwards compatibility should be considered.

**Comment by robUx4** [comment]
In general this pull request mixes formatting and actually spec changes/additions/clarifications.

I think we should go first with formatting only. Then we can discuss the actual wording of things. In general cosmetic changes should never be mixed with content changes.

File: specification.markdown [hunk 11]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
+with the width descriptor and the marker makes up one or more complete
+bytes. The tail data is as many bytes as there were zeroes in the width
+descriptor, i.e. width-1.
+
+    VINT           = VINT_WIDTH VINT_MARKER VINT_DATA
+    VINT_WIDTH     = *%b0
+    VINT_MARKER    = %b1
+    VINT_DATA      = VINT_ALIGNMENT VINT_TAIL
+    VINT_ALIGNMENT = *BIT
+    VINT_TAIL      = *BYTE
```
Comment (robUx4) [inline specification.markdown]: How can this be a BYTE when the size is 1 octet ?

File: specification.markdown [hunk 12]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
+consists of alignment data and tail data. The alignment data together
+with the width descriptor and the marker makes up one or more complete
+bytes. The tail data is as many bytes as there were zeroes in the width
+descriptor, i.e. width-1.
+
+    VINT           = VINT_WIDTH VINT_MARKER VINT_DATA
+    VINT_WIDTH     = *%b0
+    VINT_MARKER    = %b1
+    VINT_DATA      = VINT_ALIGNMENT VINT_TAIL
+    VINT_ALIGNMENT = *BIT
```
Comment (robUx4) [inline specification.markdown]: Isn't the alignment (padding?) just `*%b0` ?

File: specification.markdown [hunk 13]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
+
+Some examples of the encoding of integers of width 1 to 4. The x:es
+represent bits where the actual integer value would be stored.
+
+Width | Size | Representation
+------|------|----------------------------------------
+1     | 2^7  | 1xxx xxxx
+2     | 2^14 | 01xx xxxx xxxx xxxx
+3     | 2^21 | 001x xxxx xxxx xxxx xxxx xxxx
+4     | 2^28 | 0001 xxxx xxxx xxxx xxxx xxxx xxxx xxxx
```
Comment (robUx4) [inline specification.markdown]: You may set the alignment as a set of 0

File: specification.markdown [hunk 14]
```diff
@@ -97,242 +170,298 @@ followed by the Data Size, and then the non-interpreted Binary itself:
...
+
+## Elements semantic
+
+### Element Template
+
+Element Name:
+
+    Level:
```
Comment (robUx4) [inline specification.markdown]: Explain that a + means it's nested. It may also be another line:
`Nested: Yes/No`

File: specification.markdown (hunk 4 above)
Comment (dericed) [inline specification.markdown]: In the Element Template, we could add a value called "Element Status", then for each signature value, we can set it to "Deprecated since EBML version 1". Any better idea?

File: specification.markdown (hunk 5 above)
Comment (dericed) [inline specification.markdown]: fixed in https://github.com/MediaArea/ebml-specification/commit/db1042fb059c14a34578c025462bc90c973e7a95

**Comment by dericed** [comment]
@mbunkus I'm going to read/respond/commit to the rest of the comments from @robUx4 and you, then will skim the commit list for squash opportunities, feel free to point them out :)

File: specification.markdown (hunk 6 above)
Comment (dericed) [inline specification.markdown]: [RFC 2223](http://tools.ietf.org/html/rfc2223#section-3) states:
    
     Each line must be limited to 72 characters followed by carriage return and line feed.

but even at 72 there still isn't a reason for a line break here, fixed in https://github.com/MediaArea/ebml-specification/commit/26b506c69a22e814ce9bbfa600a3ffb71113054c

File: specification.markdown (hunk 7 above)
Comment (dericed) [inline specification.markdown]: Add an issue for now, back to this later: https://github.com/Matroska-Org/ebml-specification/issues/7

File: specification.markdown (hunk 8 above)
Comment (dericed) [inline specification.markdown]: @mbunkus, are you ok if 'Class ID' changes to 'Class Name' here as well?

    The leading bits of the Class IDs are used to identify the
        length of the Element ID. The number of leading 0's + 1 is the
        length of the ID in octets. We will refer to the leading bits
        as the Length Descriptor.

File: specification.markdown (hunk 9 above)
Comment (dericed) [inline specification.markdown]: So to clarify, I should remove this commit from the PR? Or shorten it to remove the 'shortest encoding form' requirement?

**Comment by dericed** [comment]
@robUx4 agreed about the mix of formatting and change, but not certain of the best strategy to split one PR into 2

**Comment by robUx4** [comment]
@dericed I'm afraid you have to make 2 PRs. Whichever requires the less discussion should come first so we can move forward quickly. So that would be formatting. After that each *content* change can be discussed in individual PRs that can be merged independently. Especially if the changes don't occur in the same lines.

File: specification.markdown (hunk 4 above)
Comment (robUx4) [inline specification.markdown]: Sounds okay to me.

File: specification.markdown (hunk 9 above)
Comment (robUx4) [inline specification.markdown]: Yes, it should go in its own PR.

**Comment by dericed** [comment]
@robUx4, np, I'd rather do it, right. I'll send start to send smaller sectional pull requests, first content, then formatting.

**Comment by dericed** [comment]
Update: I've been squashing the commit history here, so it's down to 30 commits from 44 or so. This PR is not yet complete. A formatting-only PR here https://github.com/Matroska-Org/ebml-specification/pull/8 should be considered before this one.

**Comment by robUx4** [comment]
Formatting PR merged

**Comment by dericed** [comment]
few more commits merged into https://github.com/Matroska-Org/ebml-specification/pull/9

**Comment by dericed** [comment]
two more pull requests based on the Element ID and Element Data Size are at https://github.com/Matroska-Org/ebml-specification/pull/10 and https://github.com/Matroska-Org/ebml-specification/pull/11
